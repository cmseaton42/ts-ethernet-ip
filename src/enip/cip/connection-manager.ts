/**************************************************
 * Original author of this file is GitHub@Penlane *
 **************************************************/

import { generateEncodedTimeout } from './util';

// Redundant Owner (Vol.1 - Table 3-5.8 Field 15)
export enum Owner {
    EXCLUSIVE = 0,
    MULTIPLE = 1,
}

// Connection Priority (Vol.1 - Table 3-5.8 Field 14,13)
export enum ConnectionType {
    NULL = 0,
    MULTICAST = 1,
    POINT_TO_POINT = 2,
    RESERVED = 3,
}

// Connection Priority (Vol.1 - Table 3-5.8 Field 11,10)
export enum Priority {
    LOW = 0,
    HIGH = 1,
    SCHEDULED = 2,
    URGENT = 3,
}

// Fixed or variable parameter (Vol.1 - Table 3-5.8 Field 9)
export enum FixedVar {
    FIXED = 0,
    VARIABLE = 1,
}

// Time Tick Value (Vol.1 - Table 3-5.11)
export enum TimePerTick {
    '1ms' = 0b0000,
    '2ms' = 0b0001,
    '4ms' = 0b0010,
    '8ms' = 0b0011,
    '16ms' = 0b0100,
    '32ms' = 0b0101,
    '64ms' = 0b0110,
    '128ms' = 0b0111,
    '256ms' = 0b1000,
    '512ms' = 0b1001,
    '1024ms' = 0b1010,
    '2048ms' = 0b1011,
    '4096ms' = 0b1100,
    '8192ms' = 0b1101,
    '16384ms' = 0b1110,
    '32768ms' = 0b1111,
}

// Timeout Multiplier (Vol.1 - Table 3-5.12)
export enum TimeoutMultiplier {
    'x4' = 0,
    'x8' = 1,
    'x16' = 2,
    'x32' = 3,
    'x64' = 4,
    'x128' = 5,
    'x256' = 6,
    'x512' = 7,
}

// Define useful constants
const ORIG_VENDOR_ID = 0x3333;
const ORIG_SERIAL_NUM = 0x1337; // :sunglasses: :rocket:
const CONN_SERIAL_NUM = 0x4242; // Jackie Robinson + The life the universe and everything :smile:

export abstract class ConnectionManager {
    /**
     * Build for Object specific connection parameters (Vol.1 - Table 3-5.8)
     */
    public static buildConnParameters(
        owner: Owner,
        type: ConnectionType,
        priority: Priority,
        fixedVar: FixedVar,
        size: number,
    ): number {
        if (owner !== 0 && owner !== 1) throw new Error('Owner can only be exclusive (0) or multiple (1)');
        if (fixedVar !== 0 && fixedVar !== 1) throw new Error('Fixedvar can only be Fixed(0) or VariableI(1)');
        if (size > 10000 || size <= 1) throw new Error('Size must be a positive number between 1 and 10000');

        if (type > 3 || type < 0) {
            throw new Error('Type can only be Null(0), Multicast(1), PointToPoint(2) or Reserved(3)');
        }

        if (priority > 3 || priority < 0) {
            throw new Error('Priority can only be Low(0), High(1), Scheduled(2) or Urgent(3)');
        }

        return (owner << 15) | (type << 13) | (priority << 10) | (fixedVar << 9) | size;
    }

    /**
     * Builds the data portion of a forwardOpen packet (Vol.1 - Table 3-5.16)
     *
     * @param {number} [RPI=10000] - Request packet interval in microseconds.
     */
    public static buildForwardOpen(
        RPI: number = 10000,
        netConnParams: number = 0x43f4, // Determined for most EIP connections
        timeOutMs: number = 2000,
        timeOutMult: TimeoutMultiplier = TimeoutMultiplier.x32,
        connSerialNum: number = CONN_SERIAL_NUM, // TODO: Make this unique
        vendorOrig: number = ORIG_VENDOR_ID,
    ): Buffer {
        if (timeOutMs <= 999) throw new Error('Timeouts Must be Positive Integers at least 1000');
        if (RPI <= 9999) throw new Error('RPI should be at least 10000us (10ms)');

        if (!(timeOutMult in TimeoutMultiplier)) {
            throw new Error('Timeout Multiplier must be a number and a multiple of 4');
        }

        const buf = Buffer.alloc(35); // Normal forward open request
        const timeout = generateEncodedTimeout(timeOutMs);

        buf.writeUInt8(timeout.timeTick, 0); // Priority / TimePerTick
        buf.writeUInt8(timeout.ticks, 1); // Timeout Ticks
        buf.writeUInt32LE(0x11111111, 2); // O->T Connection ID
        buf.writeUInt32LE(0x22222222, 6); // T->O Connection ID
        buf.writeUInt16LE(connSerialNum, 10); // Connection Serial Number
        buf.writeUInt16LE(vendorOrig, 12); // Originator VendorID
        buf.writeUInt32LE(ORIG_SERIAL_NUM, 14); // Originator Serial Number
        buf.writeUInt32LE(timeOutMult, 18); // TimeOut Multiplier
        buf.writeUInt32LE(RPI, 22); // O->T RPI
        buf.writeUInt16LE(netConnParams, 26); // O->T Network Connection Params
        buf.writeUInt32LE(RPI, 28); // T->O RPI
        buf.writeUInt16LE(netConnParams, 32); // T->O Network Connection Params
        buf.writeUInt8(0xa3, 34); // TransportClass_Trigger (Vol.1 - 3-4.4.3) -> Target is a Server, Application object of Transport Class 3.

        return buf;
    }

    /**
     * Builds the data portion of a forwardClose packet (Vol.1 - Table 3-5.19)
     *
     */
    public static buildForwardClose(
        connSerialNum: number = CONN_SERIAL_NUM, // TODO: Make this unique
        timeOutMs: number = 2000,
        vendorOrig: number = ORIG_VENDOR_ID,
        serialOrig: number = ORIG_SERIAL_NUM,
    ) {
        if (timeOutMs <= 999) throw new Error('Timeouts Must be Positive Integers at least 1000');

        const buf = Buffer.alloc(10);
        const timeout = generateEncodedTimeout(timeOutMs);

        buf.writeUInt8(timeout.timeTick, 0); // Priority / TimePerTick
        buf.writeUInt8(timeout.ticks, 1); // Timeout Ticks
        buf.writeUInt16LE(connSerialNum, 2); // Connection Serial Number
        buf.writeUInt16LE(vendorOrig, 4); // Originator VendorID
        buf.writeUInt32LE(serialOrig, 6); // Originator Serial Number

        return buf;
    }
}
