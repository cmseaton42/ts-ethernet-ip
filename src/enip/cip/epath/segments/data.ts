export enum Types {
    SIMPLE = 0x80,
    ANSI_EXTD = 0x91,
}

export enum ElementTypes {
    UINT8 = 0x28,
    UINT16 = 0x29,
    UINT32 = 0x2a,
}

/**
 * Builds EPATH Data Segment
 *
 */
export const build = (data: number | string, ANSI = true) => {
    // Build Element Segment If Int
    if (typeof data === 'number') return elementBuild(data);

    // Build symbolic segment by default
    return symbolicBuild(data, ANSI);
};

/**
 * Builds EPATH Symbolic Segment
 *
 */
export const symbolicBuild = (data: string, ANSI: boolean = true): Buffer => {
    // Initialize Buffer
    let buf = Buffer.alloc(2);

    // Write Appropriate Segment Byte
    buf.writeUInt8(ANSI ? Types.ANSI_EXTD : Types.SIMPLE, 0);

    // Write Appropriate Length
    buf.writeUInt8(ANSI ? data.length : Math.ceil(data.length / 2), 1);

    // Append Data
    buf = Buffer.concat([buf, Buffer.from(data)]);

    // Add Pad Byte if Odd Length
    if (buf.length % 2 === 1) buf = Buffer.concat([buf, Buffer.alloc(1)]); // Pad Odd Length Strings

    return buf;
};

/**
 * Builds EPATH Element Segment
 *
 */
export const elementBuild = (data: number): Buffer => {
    // Get Element Length - Data Access 2 - IOI Segments - Element Segments
    let type: ElementTypes;
    let dataBuf: Buffer;

    if (data < 256) {
        type = ElementTypes.UINT8; // UNIT8 x28 xx
        dataBuf = Buffer.alloc(1);
        dataBuf.writeUInt8(data, 0);
    } else if (data < 65536) {
        type = ElementTypes.UINT16; // UINT16 x29 00 xx xx
        dataBuf = Buffer.alloc(3);
        dataBuf.writeUInt16LE(data, 1);
    } else {
        type = ElementTypes.UINT32; // UINT32 x2a 00 xx xx xx xx
        dataBuf = Buffer.alloc(5);
        dataBuf.writeUInt32LE(data, 1);
    }

    // Initialize Buffer
    let buf = Buffer.alloc(1);

    // Write Appropriate Segment Byte
    buf.writeUInt8(type, 0);

    // Append Data
    buf = Buffer.concat([buf, dataBuf]);

    return buf;
};
