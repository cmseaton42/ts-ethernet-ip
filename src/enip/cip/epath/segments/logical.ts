const LOGICAL_SEGMENT = 1 << 5;

export enum Types {
    CLASS_ID = 0 << 2,
    INSTANCE_ID = 1 << 2,
    MEMBER_ID = 2 << 2,
    CONN_POINT = 3 << 2,
    ATTRIBUTE_ID = 4 << 2,
    SPECIAL = 5 << 2,
    SERVICE_ID = 6 << 2,
}

/**
 * Builds Single Logical Segment Buffer
 *
 * @param {Types} type - Valid Logical Segment Type
 * @param {number} address - Logical Segment Address
 * @param {boolean} [padded=true] - Padded or Packed EPATH format
 */
export const build = (type: Types, address: number, padded = true) => {
    const isValid = type in Types;
    if (!isValid) throw new Error('Invalid Logical Type Code Passed to Segment Builder');

    if (address <= 0) throw new Error('Passed Address Must be a Positive Integer');

    let buf: Buffer | null = null; // Initialize Output Buffer

    // Determine Size of Logical Segment Value and Build Buffer
    let format: number | null = null;
    if (address <= 255) {
        format = 0;

        buf = Buffer.alloc(2);
        buf.writeUInt8(address, 1);
    } else if (address > 255 && address <= 65535) {
        format = 1;

        if (padded) {
            buf = Buffer.alloc(4);
            buf.writeUInt16LE(address, 2);
        } else {
            buf = Buffer.alloc(3);
            buf.writeUInt16LE(address, 1);
        }
    } else {
        format = 2;

        if (padded) {
            buf = Buffer.alloc(6);
            buf.writeUInt32LE(address, 2);
        } else {
            buf = Buffer.alloc(5);
            buf.writeUInt32LE(address, 1);
        }
    }

    // Build Segment Byte
    const segmentByte = LOGICAL_SEGMENT | type | format;
    buf.writeUInt8(segmentByte, 0);

    return buf;
};
