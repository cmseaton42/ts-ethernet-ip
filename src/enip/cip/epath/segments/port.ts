const PORT_SEGMENT = 0 << 5;

/**
 * Builds Port Segement for EPATH
 *
 * @param {number} port - Port to leave Current Node (1 if Backplane)
 * @param {number|string} link - link address to route packet
 * @returns {buffer} EPATH Port Segment
 */
export const build = (port: number, link: number | string): Buffer => {
    if (port <= 0) throw new Error('Port Number must be a Positive Integer');
    if (typeof link === 'number' && link < 0) throw new Error('Link Number must be a Positive Integer or String');

    let buf: Buffer | null = null;
    let portIdentifierByte = PORT_SEGMENT; // Set High Byte of Segement (0x00)

    // Check Link Buffer Length
    let linkBuf: Buffer = Buffer.from([]);

    /* eslint-disable indent */
    switch (typeof link) {
        case 'string':
            linkBuf = Buffer.from(link);
            break;
        case 'number':
            linkBuf = Buffer.from([link]);
            break;
    }
    /* eslint-enable indent */

    // Build Port Buffer
    if (port < 15) {
        portIdentifierByte |= port;

        if (linkBuf.length > 1) {
            portIdentifierByte |= 0x10; // Set Flag to Identify a link of greater than 1 Byte
            buf = Buffer.alloc(2);
            buf.writeInt8(linkBuf.length, 1);
        } else {
            buf = Buffer.alloc(1);
        }
    } else {
        portIdentifierByte |= 0x0f;

        if (linkBuf.length > 1) {
            portIdentifierByte |= 0x10; // Set Flag to Identify a link of greater than 1 Byte
            buf = Buffer.alloc(4);
            buf.writeUInt8(linkBuf.length, 1);
            buf.writeUInt16LE(port, 2);
        } else {
            buf = Buffer.alloc(3);
            buf.writeUInt16LE(port, 1);
        }
    }

    buf.writeUInt8(portIdentifierByte, 0);

    // Add Link to Buffer
    buf = Buffer.concat([buf, linkBuf]); // Buffer.from(linkBuf));Buffer.alloc(1))
    return buf.length % 2 === 1 ? Buffer.concat([buf, Buffer.alloc(1)]) : buf;
};

module.exports = { build };
