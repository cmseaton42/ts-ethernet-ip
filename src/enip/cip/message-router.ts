export enum Services {
    GET_ATTRIBUTE_ALL = 0x01,
    GET_ATTRIBUTE_SINGLE = 0x0e,
    RESET = 0x05,
    START = 0x06,
    STOP = 0x07,
    CREATE = 0x08,
    DELETE = 0x09,
    MULTIPLE_SERVICE_PACKET = 0x0a,
    APPLY_ATTRIBUTES = 0x0d,
    SET_ATTRIBUTE_SINGLE = 0x10,
    FIND_NEXT = 0x11,
    READ_TAG = 0x4c,
    WRITE_TAG = 0x4d,
    READ_TAG_FRAGMENTED = 0x52,
    WRITE_TAG_FRAGMENTED = 0x53,
    READ_MODIFY_WRITE_TAG = 0x4e,
    FORWARD_OPEN = 0x54,
    FORWARD_CLOSE = 0x4e,
}

export interface IMessageData {
    service: number;
    generalStatusCode: number; // General Status Code (Vol 1 - Appendix B)
    extendedStatusLength: number;
    extendedStatus: number[] | null;
    data: Buffer | null;
}

export abstract class MessageRouter {
    /**
     * Builds a Message Router Request Buffer
     *
     * @param {Buffer} path - CIP Padded EPATH (Vol 1 - Appendix C)
     */
    public static build(service: Services, path: Buffer, data: Buffer): Buffer {
        const pathBuf = Buffer.from(path);
        const dataBuf = Buffer.from(data);

        const pathLen = Math.ceil(pathBuf.length / 2);
        const buf = Buffer.alloc(2 + pathLen * 2 + dataBuf.length);

        buf.writeUInt8(service, 0); // Write Service Code to Buffer <USINT>
        buf.writeUInt8(pathLen, 1); // Write Length of EPATH (16 bit word length)

        pathBuf.copy(buf, 2); // Write EPATH to Buffer
        dataBuf.copy(buf, 2 + pathLen * 2); // Write Service Data to Buffer

        return buf;
    }

    /**
     * Parses a Message Router Request Buffer
     *
     */
    public static parse(buf: Buffer): IMessageData {
        const decoded: IMessageData = {
            service: buf.readUInt8(0),
            generalStatusCode: buf.readUInt8(2),
            extendedStatusLength: buf.readUInt8(3),
            extendedStatus: null,
            data: null,
        };

        // Build Extended Status Array
        const arr = [];
        for (let i = 0; i < decoded.extendedStatusLength; i++) {
            arr.push(buf.readUInt16LE(i * 2 + 4));
        }
        decoded.extendedStatus = arr;

        // Get Starting Point of Message Router Data
        const dataStart = decoded.extendedStatusLength * 2 + 4;

        // Initialize Message Router Data Buffer
        const data = Buffer.alloc(buf.length - dataStart);

        // Copy Data to Message Router Data Buffer
        buf.copy(data, 0, dataStart);
        decoded.data = data;

        return decoded;
    }
}
