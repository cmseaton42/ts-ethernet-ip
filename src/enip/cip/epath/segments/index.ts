import { build as DATA } from './data';
import { build as LOGICAL } from './logical';
import { build as PORT } from './port';

enum Types {
    PORT = 0 << 5, // Communication Port to Leave Node (Shall be 1 for a Backplane), Link Address of Next Device
    LOGICAL = 1 << 5,
    NETWORK = 2 << 5,
    SYMBOLIC = 3 << 5,
    DATA = 4 << 5,
    DATATYPE_1 = 5 << 5,
    DATATYPE_2 = 6 << 6,
}

export const segments = {
    Types,
    DATA,
    LOGICAL,
    PORT,
};
