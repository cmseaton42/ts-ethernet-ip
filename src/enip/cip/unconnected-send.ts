import { MessageRouter } from './message-router';
import { segments } from './epath';
import { generateEncodedTimeout } from './util';
const { LOGICAL } = segments;

const UNCONNECTED_SEND_SERVICE = 0x52;
const UNCONNECTED_SEND_PATH = Buffer.concat([
    LOGICAL.build(LOGICAL.Types.CLASS_ID, 0x06),
    LOGICAL.build(LOGICAL.Types.INSTANCE_ID, 1),
]);

/**
 * Builds an Unconnected Send Packet Buffer
 *
 * @param {buffer} message_request - Message Request Encoded Buffer
 * @param {buffer} path - Padded EPATH Buffer
 * @param {number} [timeout=2000] - timeout (ms)
 */
export const build = (message_request: Buffer, path: Buffer, timeout: number = 2000): Buffer => {
    if (timeout < 100) timeout = 1000;

    // Get Encoded Timeout
    const encTimeout = generateEncodedTimeout(timeout);

    // Instantiate Buffer
    let buf = Buffer.alloc(2);

    // Write Encoded Timeout to Output Buffer
    buf.writeUInt8(encTimeout.time_tick, 0);
    buf.writeUInt8(encTimeout.ticks, 1);

    // Build Message Request Buffer
    const msgReqLen = message_request.length;
    const msgReqLenBuf = Buffer.alloc(2);
    msgReqLenBuf.writeUInt16LE(msgReqLen, 0);

    // Build Path Buffer
    const pathLen = Math.ceil(path.length / 2);
    const pathLenBuf = Buffer.alloc(1);
    pathLenBuf.writeUInt8(pathLen, 0);

    // Build Null Buffer
    const nullBuf = Buffer.alloc(1);

    // Assemble Unconnected Send Buffer
    if (msgReqLen % 2 === 1) {
        // requires Pad Byte after Message Request
        buf = Buffer.concat([buf, msgReqLenBuf, message_request, nullBuf, pathLenBuf, nullBuf, path]);
    } else {
        buf = Buffer.concat([buf, msgReqLenBuf, message_request, pathLenBuf, nullBuf, path]);
    }

    return MessageRouter.build(UNCONNECTED_SEND_SERVICE, UNCONNECTED_SEND_PATH, buf);
};
