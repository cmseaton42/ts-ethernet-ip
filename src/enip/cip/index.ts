import * as EPATH from './epath';
import { Types as DataTypes } from './data-types';
import { MessageRouter, Services } from './message-router';
import * as UnconnectedSend from './unconnected-send';
import * as util from './util';
import {
    ConnectionManager,
    Owner,
    ConnectionType,
    Priority,
    FixedVar,
    TimePerTick,
    TimeoutMultiplier,
} from './connection-manager';

const connManagerEnums = {
    Owner,
    ConnectionType,
    Priority,
    FixedVar,
    TimePerTick,
    TimeoutMultiplier,
};

export const cip = {
    EPATH,
    MessageRouter,
    Services,
    DataTypes,
    UnconnectedSend,
    util,
    ConnectionManager,
    connManagerEnums,
};
