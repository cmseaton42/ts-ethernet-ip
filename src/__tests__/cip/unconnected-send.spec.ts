import { build, generateEncodedTimeout } from '../../enip/cip/unconnected-send';
import { MessageRouter } from '../../enip/cip/message-router';
import { segments } from '../../enip/cip/epath/segments';
const { PORT } = segments;

describe('Unconnected Send Service', () => {
    describe('Timeout Encoding Utility', () => {
        it('Generates Appropriate Outputs', () => {
            const fn = (arg: number) => generateEncodedTimeout(arg);

            expect(fn(2304)).toMatchObject({ time_tick: 8, ticks: 9 });
            expect(fn(2400)).toMatchObject({ time_tick: 5, ticks: 75 });
            expect(fn(2000)).toMatchObject({ time_tick: 4, ticks: 125 });
            expect(() => fn(-1)).toThrow();
        });
    });

    describe('Message Build Utility', () => {
        it('Generates Appropriate Output', () => {
            const readTag_Path1 = Buffer.from('sometag');
            const readTag_Path2 = Buffer.from('sometag2');
            const readTag_Data = Buffer.alloc(2);
            readTag_Data.writeUInt16LE(1, 0);

            const MR1 = MessageRouter.build(0x4c, readTag_Path1, readTag_Data);
            const MR2 = Buffer.from('odd length!'); // odd length buffer

            let test1 = build(MR1, PORT.build(1, 5));
            expect(test1).toMatchSnapshot();

            let test2 = build(MR2, PORT.build(1, 5), 50);
            expect(test2).toMatchSnapshot();
        });
    });
});
