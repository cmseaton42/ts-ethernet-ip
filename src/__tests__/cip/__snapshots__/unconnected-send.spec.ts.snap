// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unconnected Send Service Message Build Utility Generates Appropriate Output 1`] = `
Object {
  "data": Array [
    82,
    2,
    32,
    6,
    36,
    1,
    4,
    125,
    12,
    0,
    76,
    4,
    115,
    111,
    109,
    101,
    116,
    97,
    103,
    0,
    1,
    0,
    1,
    0,
    1,
    5,
  ],
  "type": "Buffer",
}
`;

exports[`Unconnected Send Service Message Build Utility Generates Appropriate Output 2`] = `
Object {
  "data": Array [
    82,
    2,
    32,
    6,
    36,
    1,
    3,
    125,
    11,
    0,
    111,
    100,
    100,
    32,
    108,
    101,
    110,
    103,
    116,
    104,
    33,
    0,
    1,
    0,
    1,
    5,
  ],
  "type": "Buffer",
}
`;
